File: app/src/main/java/com/nfcalarmclock/NacActiveAlarmNotification.java
package com.nfcalarmclock;

import android.annotation.TargetApi;
import android.app.Notification;
import android.app.NotificationChannel;
import android.app.PendingIntent;
import android.content.Context;
import android.content.Intent;
import android.content.res.Resources;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.os.Build;
import androidx.core.app.NotificationCompat;
import androidx.core.app.NotificationManagerCompat;
import java.lang.Float;
import java.util.Calendar;
import java.util.Locale;

/**
 * Notification to display for active alarms.
 */
public class NacActiveAlarmNotification
	extends NacNotification
{

	/**
	 * Group key.
	 */
	public static final String GROUP = "NacNotiGroupActiveAlarm";

	/**
	 * Alarm.
	 */
	private NacAlarm mAlarm;

	/**
	 */
	public NacActiveAlarmNotification()
	{
		super();
		this.mAlarm = null;
	}

	/**
	 */
	public NacActiveAlarmNotification(Context context)
	{
		super(context);
		this.mAlarm = null;
	}

	/**
	 * @see builder
	 */
	@Override
	protected NotificationCompat.Builder builder()
	{
		Context context = this.getContext();
		Bitmap icon = this.getIcon();
		PendingIntent activityPending = this.getContentPendingIntent();
		PendingIntent snoozePending = this.getSnoozePendingIntent();
		PendingIntent dismissPending = this.getDismissPendingIntent(activityPending);
		NacSharedConstants cons = new NacSharedConstants(context);

		return super.builder()
			.setLargeIcon(icon)
			.setFullScreenIntent(activityPending, true)
			.setAutoCancel(false)
			.setOngoing(true)
			.setShowWhen(true)
			.setTicker(cons.getAppName())
			.addAction(R.mipmap.snooze, cons.getActionSnooze(), snoozePending)
			.addAction(R.mipmap.dismiss, cons.getActionDismiss(), dismissPending);
	}

	/**
	 */
	@TargetApi(Build.VERSION_CODES.O)
	@Override
	protected NotificationChannel createChannel()
	{
		NotificationChannel channel = super.createChannel();
		if (channel == null)
		{
			return null;
		}

		channel.setShowBadge(true);
		channel.enableLights(true);
		channel.enableVibration(true);
		return channel;
	}

	/**
	 * @return The alarm.
	 */
	public NacAlarm getAlarm()
	{
		return this.mAlarm;
	}

	/**
	 * @return The notification description.
	 */
	protected String getChannelDescription()
	{
		Context context = this.getContext();
		NacSharedConstants cons = new NacSharedConstants(context);
		return cons.getDescriptionActiveNotification();
	}

	/**
	 * @return The notification name.
	 */
	protected String getChannelName()
	{
		Context context = this.getContext();
		NacSharedConstants cons = new NacSharedConstants(context);
		return cons.getActiveNotification();
	}

	/**
	 * @see getChannel
	 */
	protected String getChannelId()
	{
		return "NacNotiChannelActiveAlarm";
	}

	/**
	 * @see getContentPendingIntent
	 */
	protected PendingIntent getContentPendingIntent()
	{
		Context context = this.getContext();
		NacAlarm alarm = this.getAlarm();
		int id = (alarm == null) ? 0 : alarm.getId();
		Intent intent = NacIntent.createAlarmActivity(context, alarm);

		return PendingIntent.getActivity(context, id, intent,
			PendingIntent.FLAG_UPDATE_CURRENT);
	}

	/**
	 * @see getContentText
	 */
	public String getContentText()
	{
		Context context = this.getContext();
		NacAlarm alarm = this.getAlarm();
		Calendar now = Calendar.getInstance();
		String time = NacCalendar.Time.getFullTime(context, now);
		String name = alarm.getName();
		Locale locale = Locale.getDefault();

		return name.isEmpty() ? time
			: String.format(locale, "%1$s  —  %2$s", time, name);
	}

	/**
	 * @return The pending intent to use when dismissing the alarm.
	 */
	private PendingIntent getDismissPendingIntent(
		PendingIntent activityPendingIntent)
	{
		Context context = this.getContext();
		NacAlarm alarm = this.getAlarm();

		if (NacNfc.shouldUseNfc(context, alarm))
		{
			return activityPendingIntent;
		}

		Intent intent = NacIntent.dismissForegroundService(context, alarm);
		return PendingIntent.getService(context, 0, intent, 0);
	}

	/**
	 * @see getGroup
	 */
	protected String getGroup()
	{
		return GROUP;
	}

	/**
	 * @return The notification large icon.
	 */
	protected Bitmap getIcon()
	{
		Context context = this.getContext();
		Resources res = context.getResources();
		Bitmap icon = BitmapFactory.decodeResource(res, R.mipmap.app);
		float density = res.getDisplayMetrics().density;
		float size = 0f;

		if (Float.compare(density, 4.0f) >= 0)
		{
			size = 256f;
		}
		else if (Float.compare(density, 3.0f) >= 0)
		{
			size = 192f;
		}
		else if (Float.compare(density, 2.0f) >= 0)
		{
			size = 128f;
		}
		else if (Float.compare(density, 1.5f) >= 0)
		{
			size = 96f;
		}
		else if (Float.compare(density, 1.0f) >= 0)
		{
			size = 64f;
		}
		else if (Float.compare(density, 0.75f) >= 0)
		{
			size = 48f;
		}
		else
		{
			return icon;
		}

		return Bitmap.createScaledBitmap(icon, (int)(size*density),
			(int)(size*density), true);
	}

	/**
	 * @see getId
	 */
	protected int getId()
	{
		return 79;
	}

	/**
	 * @see getImportance
	 */
	protected int getImportance()
	{
		return NotificationManagerCompat.IMPORTANCE_HIGH;
	}

	/**
	 * @see getPriority
	 */
	protected int getPriority()
	{
		return NotificationCompat.PRIORITY_MAX;
	}

	/**
	 * @return The notification title.
	 */
	public String getTitle()
	{
		Context context = this.getContext();
		Locale locale = Locale.getDefault();
		NacSharedConstants cons = new NacSharedConstants(context);

		return String.format(locale, "<b>%s</b>", cons.getAppName());
	}

	/**
	 * @return The pending intent to use when snoozing.
	 */
	private PendingIntent getSnoozePendingIntent()
	{
		Context context = this.getContext();
		Intent intent = new Intent(NacForegroundService.ACTION_SNOOZE_ALARM,
			null, context, NacForegroundService.class);

		return PendingIntent.getService(context, 0, intent, 0);
	}

	/**
	 * @return True if the given group matches this notification group.
	 */
	public static boolean isActiveGroup(String groupKey)
	{
		return GROUP.equals(groupKey);
	}

	/**
	 * Set the alarm.
	 */
	public void setAlarm(NacAlarm alarm)
	{
		this.mAlarm = alarm;
	}

}

File: app/src/main/kotlin/com/nfcalarmclock/alarm/activealarm/NacActiveAlarmNotification.kt
package com.nfcalarmclock.alarm.activealarm

import android.app.NotificationChannel
import android.app.PendingIntent
import android.content.Context
import android.os.Build
import androidx.annotation.OptIn
import androidx.annotation.RequiresApi
import androidx.core.app.NotificationCompat
import androidx.core.app.NotificationManagerCompat
import androidx.core.content.ContextCompat
import androidx.media3.common.util.UnstableApi
import com.nfcalarmclock.R
import com.nfcalarmclock.alarm.db.NacAlarm
import com.nfcalarmclock.alarm.options.nfc.NacNfc
import com.nfcalarmclock.shared.NacSharedPreferences
import com.nfcalarmclock.system.NacCalendar
import com.nfcalarmclock.view.notification.NacBaseNotificationBuilder
import com.nfcalarmclock.view.toSpannedString
import java.util.Calendar

/**
 * Active alarm notification.
 *
 * @param context Context.
 * @param alarm Alarm.
 */
class NacActiveAlarmNotification(
	context: Context,
	private val alarm: NacAlarm?
) : NacBaseNotificationBuilder(context, "NacNotiChannelActiveAlarm")
{

	/**
	 * @see NacBaseNotificationBuilder.id
	 */
	override val id: Int
		get() = ID

	/**
	 * @see NacBaseNotificationBuilder.channelName
	 */
	override val channelName: String
		get() = context.getString(R.string.title_active_alarms)

	/**
	 * @see NacBaseNotificationBuilder.channelDescription
	 */
	override val channelDescription: String
		get() = context.getString(R.string.description_active_alarm)

	/**
	 * @see NacBaseNotificationBuilder.channelImportance
	 */
	override val channelImportance: Int = NotificationManagerCompat.IMPORTANCE_HIGH

	/**
	 * @see NacBaseNotificationBuilder.priorityLevel
	 */
	override val priorityLevel: Int = NotificationCompat.PRIORITY_MAX

	/**
	 * @see NacBaseNotificationBuilder.group
	 */
	override val group: String = "NacNotiGroupActiveAlarm"

	/**
	 * @see NacBaseNotificationBuilder.contentText
	 */
	override val contentText: String
		get()
		{
			// Get the full time
			val now = Calendar.getInstance()
			val time = NacCalendar.getFullTime(context, now)

			// Get the alarm name
			val name = alarm?.name ?: ""

			return if (name.isEmpty())
			{
				time
			}
			else
			{
				"$time  —  $name"
			}
		}

	/**
	 * @see NacBaseNotificationBuilder.contentPendingIntent
	 */
	override val contentPendingIntent: PendingIntent
		get()
		{
			val id = alarm?.id ?: 0
			val intent = NacActiveAlarmActivity.getStartIntent(context, alarm)

			// Determine the pending intent flags
			val flags = PendingIntent.FLAG_UPDATE_CURRENT or PendingIntent.FLAG_IMMUTABLE

			// Return the pending intent for the activity
			return PendingIntent.getActivity(context, id.toInt(), intent, flags)
		}

	/**
	 * The pending intent to use when dismissing the alarm.
	 */
	private val dismissPendingIntent: PendingIntent
		@OptIn(UnstableApi::class)
		get()
		{
			// Create an intent to dismiss the active alarm service
			val intent = NacActiveAlarmService.getDismissIntent(context, alarm)

			// Determine the pending intent flags
			val flags = PendingIntent.FLAG_CANCEL_CURRENT or PendingIntent.FLAG_IMMUTABLE

			// Create the pending intent
			return PendingIntent.getService(context, 0, intent, flags)
		}

	/**
	 * The pending intent to use when snoozing.
	 */
	private val snoozePendingIntent: PendingIntent
		@OptIn(UnstableApi::class)
		get()
		{
			val intent = NacActiveAlarmService.getSnoozeIntent(context, alarm)

			// Determine the pending intent flags
			val flags = PendingIntent.FLAG_CANCEL_CURRENT or PendingIntent.FLAG_IMMUTABLE

			// Create the pending intent
			return PendingIntent.getService(context, 0, intent, flags)
		}

	/**
	 * Whether the alarm should use NFC or not.
	 */
	private val shouldUseNfc: Boolean
		get() = (alarm != null) && alarm.shouldUseNfc && NacSharedPreferences(context).shouldShowNfcButton

	/**
	 * Constructor.
	 */
	init
	{
		// Get shared preferences
		val sharedPreferences = NacSharedPreferences(context)

		// Create the channel
		setupChannel()

		// Get the title
		val appName = context.getString(R.string.app_name)

		// Build the notification
		// Note: Added the parentheses so that the custom addAction() can be called
		(this.setPriority(priorityLevel)
			.setCategory(category)
			.setGroup(group)
			.setContentTitle("<b>$appName</b>".toSpannedString())
			.setContentText(contentText)
			.setContentIntent(contentPendingIntent)
			.setSmallIcon(smallIcon)
			.setTicker(channelName)
			.setForegroundServiceBehavior(NotificationCompat.FOREGROUND_SERVICE_IMMEDIATE)
			.setColor(ContextCompat.getColor(context, R.color.ic_launcher_background))
			.setAutoCancel(false)
			.setOngoing(true)
			.setShowWhen(true)
			.setSound(null) as NacBaseNotificationBuilder)
			.addAction(R.drawable.snooze, R.string.action_alarm_snooze, snoozePendingIntent)
			.apply {
				// Check if NFC does not need to be used to dismiss the alarm
				// Note: This evaluates to False on the emulator because the emulator
				// is unable to use NFC
				if (!NacNfc.exists(context) || !shouldUseNfc)
				{
					// Add the dismiss button to the notification
					addAction(R.drawable.dismiss, R.string.action_alarm_dismiss, dismissPendingIntent)
				}

				// Check if battery saver option is disabled
				if (!sharedPreferences.shouldSaveBatteryInAlarmScreen)
				{
					setFullScreenIntent(contentPendingIntent, true)
				}

			}
	}

	/**
	 * @see NacBaseNotificationBuilder.createChannel
	 */
	@RequiresApi(Build.VERSION_CODES.O)
	override fun createChannel(): NotificationChannel
	{
		// Create the channel
		val channel = super.createChannel()

		// Setup the channel
		channel.setShowBadge(true)
		channel.enableLights(true)
		channel.enableVibration(true)
		channel.setSound(null, null)

		return channel
	}

	companion object
	{

		/**
		 * Notification ID.
		 */
		const val ID = 69

	}

}